horovod/common/elastic.py

```diff
@@ -0,0 +1,169 @@
+        """
+        self.save()
+        self.check_host_updates()
+
+    def check_host_updates(self):
+        """Checks that a notification has been sent indicating that hosts can be added or will be removed.
+
+        Raises a `HostsUpdatedInterrupt` if such a notification has been received.
+        """
+        updated = False
+        if self._rank() == 0:
+            # Iterate through the update messages sent from the server. If the update timestamp
+            # is greater than the last update timestamp, then trigger a HostsUpdatedException.
+            while not self._host_messages.empty():
+                timestamp = self._host_messages.get()
+                if timestamp > self._last_updated_timestamp:
+                    self._last_updated_timestamp = timestamp
+                    updated = True
+
+        # In order to ensure all workers raise the exception at the same time, we need to sync
+        # the updated state across all the workers.
+        updated = self._bcast_bool(updated)
+
+        # At this point, updated state is globally consistent across all ranks.
+        if updated:
+            raise HostsUpdatedInterrupt()
+
+    def save(self):
+        """Saves state to host memory."""
+        raise NotImplementedError()
+
+    def restore(self):
+        """Restores the last committed state, undoing any uncommitted modifications."""
+        raise NotImplementedError()
+
+    def sync(self):
+        """Synchronize state across workers."""
+        raise NotImplementedError()
+
+
+class ObjectState(State):
+    """State for simple Python objects.
+
+    Every object is specified as a keyword argument, and will be assigned as an attribute.
+
+    Args:
+        bcast_object: Horovod broadcast object function used to sync state dictionary.
+        kwargs: Properties to sync, will be exposed as attributes of the object.
+    """
+    def __init__(self, bcast_object, get_rank, **kwargs):
+        self._bcast_object = bcast_object
+        self._saved_state = kwargs
+        self._set_attrs()
+        super(ObjectState, self).__init__(bcast_bool=bcast_object, get_rank=get_rank)
+
+    def save(self):
+        new_state = {}
+        for attr in self._saved_state.keys():
+            new_state[attr] = getattr(self, attr)
+        self._saved_state = new_state
+
+    def restore(self):
+        self._set_attrs()
+
+    def sync(self):
+        if self._saved_state:
+            self._saved_state = self._bcast_object(self._saved_state)
+            self._set_attrs()
+
+    def _set_attrs(self):
+        for attr, value in self._saved_state.items():
+            setattr(self, attr, value)
+
+
+def run_fn(func, reset):
+    @functools.wraps(func)
+    def wrapper(state, *args, **kwargs):
+        notification_manager.init()
+        notification_manager.register_listener(state)
+
+        try:
+            reset_required = False
+            while True:
```

Comment:
`reset()` and `state.on_reset()` can be moved to the end of the loop, which won't be executed when `return func(...)` returns. No need for `reset_required` then.

Reply:
Much cleaner!